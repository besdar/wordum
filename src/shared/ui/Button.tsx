import React, {useState} from 'react';
import {Button as PaperButton} from 'react-native-paper';

type Props = React.ComponentProps<typeof PaperButton>;

export const Button = ({onPress, disabled, loading, ...props}: Props) => {
  const [isLoading, setIsLoading] = useState(false);

  return (
    <PaperButton
      {...props}
      onPress={e => {
        setIsLoading(true);
        Promise.resolve(onPress?.(e)).finally(() => setIsLoading(false));
      }}
      loading={loading || isLoading}
      disabled={disabled || isLoading}
    />
  );
};
