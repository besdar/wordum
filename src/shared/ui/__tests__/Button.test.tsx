import React from 'react';
import {render, screen, userEvent} from '@testing-library/react-native';
import {Button} from '../Button';

describe('Button', () => {
  it('should match the snapshot', () => {
    const {toJSON} = render(<Button>Test</Button>);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should not call onPress when disabled', async () => {
    const mockOnPress = jest.fn();

    render(
      <Button onPress={mockOnPress} disabled>
        Press Me
      </Button>,
    );

    const button = screen.getByRole('button');

    const user = userEvent.setup();
    await user.press(button);

    expect(mockOnPress).not.toHaveBeenCalled();
  });
});
