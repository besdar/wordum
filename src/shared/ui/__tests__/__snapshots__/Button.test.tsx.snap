// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match the snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "transparent",
      "borderRadius": 20,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0,
      "shadowRadius": 0,
    }
  }
  testID="button-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 20,
        "borderStyle": "solid",
        "borderWidth": 0,
        "flex": undefined,
        "minWidth": 64,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
    testID="button-container"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 20,
          },
        ]
      }
      testID="button"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "letterSpacing": 0.1,
                  "lineHeight": 20,
                },
                [
                  {
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  false,
                  {
                    "marginHorizontal": 12,
                  },
                  undefined,
                  false,
                  {
                    "color": "rgba(103, 80, 164, 1)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  undefined,
                ],
              ],
            ]
          }
          testID="button-text"
        >
          Test
        </Text>
      </View>
    </View>
  </View>
</View>
`;
